### PR TITLE
add paddle.gather for API2.0

### PR DESCRIPTION
**Pull request by wangchaochaohu**
### PR types

New features
### PR changes

OPs
### Describe

gather(x, index, axis=None, name=None)

### CODE REVIEW

File: python/paddle/fluid/layers/nn.py [hunk 1]
```diff
@@ -8168,19 +8168,21 @@ def gather(input, index, overwrite=True):
                        [5, 6]]
 
     Args:
-        input (Variable): The source input tensor with rank>=1. Supported data type is
+        input (Temspr): The source input tensor with rank>=1. Supported data type is
```
Comment (TCChenlong) [inline python/paddle/fluid/layers/nn.py]: Tensor

**Review by heavengate — APPROVED** [review]

**Review by ForFishes — APPROVED** [review]
lgtm

File: python/paddle/fluid/layers/nn.py [hunk 2]
```diff
@@ -8179,19 +8179,21 @@ def gather(input, index, overwrite=True):
                        [5, 6]]
 
     Args:
-        input (Variable): The source input tensor with rank>=1. Supported data type is
+        input (Tensor): The source input tensor with rank>=1. Supported data type is
```
Comment (XiaoguangHu01) [inline python/paddle/fluid/layers/nn.py:8182]: fluid.layer.gather加一下deprecated装饰器

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

File: python/paddle/tensor/manipulation.py [hunk 3]
```diff
@@ -679,57 +677,75 @@ def gather(input, index, overwrite=True):
...
             float16 (only for GPU).
-        index (Variable): The index input tensor with rank=1. Data type is int32 or int64.
-        overwrite (bool, optional): The mode that updating the grad when has same index.
-            If True, use the overwrite mode to update the grad of the same index,
-            if False, use the accumulate mode to update the grad of the same index.
-            Default value is True.
-
-
+        index (Tensor): The index input tensor with rank=1. Data type is int32 or int64.
+        axis (Tensor|int, optional): The axis of input to be gathered, it's can be int or a Tensor with data type is int32 or int64. Default: if None, the axis is 0.
```
Comment (TCChenlong) [inline python/paddle/tensor/manipulation.py:696]: Default: if None 语法不通顺

**Review by TCChenlong — APPROVED** [review]
LGTM

**Review by zhiqiu — APPROVED** [review]
LGTM for op_function_generator.cc

File: python/paddle/fluid/layers/nn.py (hunk 2 above)
Comment (wangchaochaohu) [reply python/paddle/fluid/layers/nn.py:8182]: https://github.com/PaddlePaddle/Paddle/pull/26579 已加

File: python/paddle/tensor/manipulation.py (hunk 3 above)
Comment (wangchaochaohu) [reply python/paddle/tensor/manipulation.py:696]: https://github.com/PaddlePaddle/Paddle/pull/26579 已改

**Pull request merged**